third_party/move/move-vm/types/src/loaded_data/runtime_access_specifier.rs

```diff
@@ -0,0 +1,536 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Runtime representation of access control specifiers.
+//!
+//! Specifiers are represented as a list of inclusion and exclusion clauses. Each
+//! of those clauses corresponds to an `acquires A`, `reads A`, or `writes A`
+//! declaration in the language. Exclusions stem from negation, e.g. `!reads A`.
+//!
+//! Specifiers support access check via `AccessSpecifier::enables`. Moreover,
+//! access specifiers can be joined via `AccessSpecifier::join`. The join of two access
+//! specifiers behaves like intersection: for `a1 join a2`, access is allowed if it
+//! is both allowed by `a1` and `a2`. Joining happens when a function is entered which
+//! has access specifiers: then the current active access specifier is joined with the
+//! function's specifier. The join operator is complete (no approximation). A further
+//! operator `AccessSpecifier::subsumes` allows to test whether one specifier
+//! allows all the access of the other. This used to abort execution if a function
+//! is entered which declares accesses not allowed by the context. However, the
+//!`subsumes` function is incomplete. This is semantically sound since
+//! if subsume is undecided, abortion only happens later at the time of actual access
+//! instead of when the function is entered.
+//!
+//! The `join` operation attempts to simplify the resulting access specifier, making
+//! access checks faster and keeping memory use low. This is only implemented for
+//! inclusions, which are fully simplified. Exclusions are accumulated.
+//! There is potential for optimization by simplifying exclusions but since those are effectively
+//! negations, such a simplification is not trivial and may require recursive specifiers, which
+//! we like to avoid.
+
+use crate::{
+    loaded_data::runtime_types::{StructIdentifier, Type},
+    values::{Reference, SignerRef, Value},
+};
+use itertools::Itertools;
+use move_binary_format::{
+    errors::{PartialVMError, PartialVMResult},
+    file_format::{AccessKind, LocalIndex},
+};
+use move_core_types::{
+    account_address::AccountAddress, language_storage::ModuleId, vm_status::StatusCode,
+};
+use std::{fmt, fmt::Debug};
+
+/// Represents an access specifier.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug)]
+pub enum AccessSpecifier {
+    /// Universal access granted
+    Any,
```

Comment:
Is `Any` the same as `Constraint(vec![], vec![])`?

Reply:
This one is actually the opposite, because the inclusions are empty. So Any is needed.